docs/development.rst

```diff
@@ -2,9 +2,177 @@
+    initializes the input and model parameters of a particular run. It
+    is also called before the adapter is used in any way and at most
+    once before a call to :func:`~openscm.adapter.Adapter.run` or
+    :func:`~openscm.adapter.Adapter.step`.
+
+    This and the
+    :func:`~openscm.adapter.Adapter._initialize_run_parameters` method
+    are separated for higher efficiency when doing ensemble runs for
+    models that have additional overhead for changing run parameters.
+
+    .. code:: python
+
+        def _initialize_model_input(self) -> None:
+            """
+            TODO Initialize model input by reading input parameters
+            from ``self._parameters`` (see below).
+            """
+
+  - The :func:`~openscm.adapter.Adapter._reset` method resets the model
+    to prepare for a new run. It is called once after each call of
+    :func:`~openscm.adapter.Adapter.run` and to reset the model after
+    several calls to :func:`~openscm.adapter.Adapter.step`.
+
+    .. code:: python
+
+        def _reset(self) -> None:
+            # TODO Reset the model
+
+  - The :func:`~openscm.adapter.Adapter._run` method runs the model
+    over the full time range (as given by the times set by the
+    previous call to
+    :func:`~openscm.adapter.Adapter._initialize_run_parameters`). You
+    should at least implement this function.
+
+    .. code:: python
+
+        def _run(self) -> None:
+            """
+            TODO Run the model and write output parameters to
+            `self._parameters` (see below).
+            """
+
+  - The :func:`~openscm.adapter.Adapter._step` method does a single
+    time step. You can get the current time from
+    :attr:`self._current_time
+    <openscm.adapter.Adapter._current_time>`, which you should
+    increase by the time step length and return its value. If your
+    model does not support stepping just do ``raise
+    NotImplementedError`` here.
+
+    .. code:: python
+
+        def _step(self) -> None:
+            """
+            TODO Do a single time step and write corresponding output
+            parameters to ``self._parameters`` (see below).
+            """
+            self._current_time += YEAR
+
+  - The :func:`~openscm.adapter.Adapter._shutdown` method cleans up the
+    adapter.
+
+    .. code:: python
+
+        def _shutdown(self) -> None:
+            # TODO Shut down model
+
+
+Reading model and input parameters
+**********************************
+
+Model parameters and input data (referred to as general "parameters"
+in OpenSCM) are **pulled** from the :func:`~openscm.core.ParameterSet`
+provided by the OpenSCM Core. From this, you read the parameters in
+the :ref:`time frame <timeframes>` and :ref:`unit <units>` as needed
+by the specific model; conversion is done interally if possible.
+
+OpenSCM defines a :ref:`set of standard parameters
+<standard-parameters>` to be shared between different SCMs. You should
+use these where appropriate.
```

Comment:
What does 'use these where appropriate' mean? When you can, write output to the standard set of OpenSCM parameters as well as to your model specific parameters?

Reply:
Yes, they should be used as input where they apply and as output where possible. I don’t know how to put that better...

Reply:
How about 'As far as possible, adapters should be able to take all of them as input and should write their values to the output `ParameterSet`.'

Reply:
nice! I added a “Conversion of specific forms of these (aggregation, unit conversion, and time frame adjustment) is ensured by the `ParameterSet` class.”